src/Functional/ZipAll.php

```diff
@@ -0,0 +1,74 @@
+<?php
+/**
+ * Copyright (C) 2011-2016 by Lars Strojny <lstrojny@php.net>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+namespace Functional;
+
+use Functional\Exceptions\InvalidArgumentException;
+use Traversable;
+
+/**
+ * Recombines arrays by index and applies a callback optionally
+ * When the input colletions are different lengths the resulting collections
+ * will all have the length which is required to fit all the keys
+ *
+ * @param $args array|Traversable $collection One or more callbacks
+ * @return array
+ */
+function zip_all(...$args)
+{
+    $callback = null;
+    if (is_callable(end($args))) {
+        $callback = array_pop($args);
+    }
+
+    foreach ($args as $position => $arg) {
+        InvalidArgumentException::assertCollection($arg, __FUNCTION__, $position + 1);
+    }
+
+    $resultKeys = [];
+    foreach ($args as $arg) {
+        foreach ($arg as $index => $value) {
+            $resultKeys[] = $index;
+        }
+    }
+
+    $resultKeys = array_unique($resultKeys);
+
+    $result = [];
+
+    foreach ($resultKeys as $key) {
+        $zipped = [];
+
+        foreach ($args as $arg) {
+            $zipped[] = isset($arg[$key]) ? $arg[$key] : null;
+        }
+
+        if ($callback !== null) {
+            /** @var callable $callback */
```

Comment:
Better to assign the type where the variable is assigned first (if type assignment is needed).